homeassistant/components/alarm_control_panel/group.py

```diff
@@ -0,0 +1,241 @@
+        yield from self._async_send_message(SERVICE_ALARM_ARM_HOME,
+                                            code=code)
+
+    @asyncio.coroutine
+    def async_alarm_arm_away(self, code=None):
+        """Send arm away command."""
+        yield from self._async_send_message(SERVICE_ALARM_ARM_AWAY,
+                                            code=code)
+
+    @asyncio.coroutine
+    def async_alarm_arm_night(self, code=None):
+        """Send arm night command."""
+        yield from self._async_send_message(SERVICE_ALARM_ARM_NIGHT,
+                                            code=code)
+
+    @asyncio.coroutine
+    def async_alarm_trigger(self, code=None):
+        """Send alarm trigger command."""
+        yield from self._async_send_message(SERVICE_ALARM_TRIGGER,
+                                            code=code)
+
+    @asyncio.coroutine
+    def async_alarm_arm_custom_bypass(self, code=None):
+        """Send arm custom bypass command."""
+        yield from self._async_send_message(SERVICE_ALARM_ARM_CUSTOM_BYPASS,
+                                            code=code)
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        # _update_state always behaves as if a transition was in progress.
+        # But if the pre- and post-state match, we're not "pending"!
+        self._update_state()
+        if self._pre_state == self._post_state:
+            self._state = self._post_state
+        else:
+            self._state = STATE_ALARM_PENDING
+        return self._state
+
+    def _update_state(self):
+        """Update the current transition state of the device."""
+        def _common_state(states):
+            # This can happen if the sub-panels have not loaded yet.
+            if len(states) == 0:
+                return STATE_ALARM_DISARMED
+
+            # One triggered sub-panel is enough to trigger the whole panel.
+            if states[STATE_ALARM_TRIGGERED]:
+                return STATE_ALARM_TRIGGERED
+
+            # This only happens if a member does not provide pre or
+            # post_pending_state.  In that case, ensure the pending state
+            # percolates up to device_state_attributes, and the attribute
+            # will be left out in the group as well.
+            if states[STATE_ALARM_PENDING]:
+                return STATE_ALARM_PENDING
+
+            # If all entities have the same state, great.  Otherwise,
+            # we can only summarize the state as "custom bypass".
+            if len(states) == 1:
+                return next(iter(states))
+            return STATE_ALARM_ARMED_CUSTOM_BYPASS
+
+        # First collect the pre- and post-states for all devices.
+        self._triggered_panels = []
+        pre_states = Counter()
+        post_states = Counter()
+        for entity_id, state in self._states:
+            if state.state == STATE_ALARM_PENDING and \
+                    (ATTR_PRE_PENDING_STATE in state.attributes):
```

Comment:
I think it's a bit weird that we depend on the state attributes to define state. Maybe we can store the previous pre/post states in two new instance attibutes, instead? Let me know if I'm missing the point here.

Reply:
The code is looking "inside" the subpanels' state attributes to understand more clearly what the subpanels are doing, and possible synthesize those same state attributes for the group (`ATTR_PRE_PENDING_STATE` goes to `self._pre_state`, `ATTR_POST_PENDING_STATE` goes to `self._post_state`.

It would certainly be possible to make before/after states into instance attributes that are available even in non-PENDING states; and in fact it may possibly make some sense for the "before" state, because that is effectively the "current operating state" of the panel (I think it makes less sense for the "after" state). However, considering that most alarm platforms do not provide PRE_PENDING_STATE, I think it's cleaner to leave it as a "nice to have" state attribute.